### PR TITLE
tuple fix

### PR DESCRIPTION
**Pull request by lockwo**
Fixes https://github.com/patrick-kidger/equinox/issues/715

### CODE REVIEW

**Comment by lockwo** [comment]
The tests will fail, but there's some issues that are unrelated: https://github.com/patrick-kidger/equinox/issues/716

File: equinox/_tree.py [hunk 1]
```diff
@@ -51,6 +51,22 @@ def count(self, item):
...
+    def __eq__(self, other):
+        if isinstance(other, _DistinctTuple):
+            return self.contents == other.contents
+        return False
+
+    def __hash__(self):
+        return hash(self.contents)
+
+    def __repr__(self):
+        return "Distinct Tuple: " + repr(self.contents)
```
Comment (patrick-kidger) [inline equinox/_tree.py:67]: I think a lot of this is unused -- just
```python
class _DistinctTuple:
    pass
```
should suffice, I think?

(In particular I don't think reprs are ever user-visible, and the `contents` are always empty.)

File: equinox/_tree.py [hunk 2]
```diff
@@ -229,7 +250,11 @@ def _make_replacement(x: _Node) -> Any:
                 "  >>> eqx.tree_at(lambda t: t[0], (None, None, 1), True,\n"
                 "  ...             is_leaf=lambda x: x is None)"
             )
-
+    out = jtu.tree_map(
+        lambda x: x if not isinstance(x, _DistinctTuple) else (),
```
Comment (patrick-kidger) [inline equinox/_tree.py:254]: Nit: both here and above, can you remove the `not` and flip the branches?

File: tests/test_tree.py [hunk 3]
```diff
@@ -31,6 +31,22 @@ def test_tree_at_replace(getkey):
...
+
+    assert pytree1[:-2] == true_pytree1[:-2]
+    assert pytree2[:-2] == true_pytree2[:-2]
+    assert jnp.all(pytree1[-2]["a"] == true_pytree1[-2]["a"])
+    assert jnp.all(pytree2[-2]["a"] == true_pytree2[-2]["a"])
+    assert jnp.all(pytree1[-1].weight == true_pytree1[-1].weight)
+    assert jnp.all(pytree1[-1].bias == true_pytree1[-1].bias)
+    assert jnp.all(pytree2[-1].weight == true_pytree2[-1].weight)
+    assert jnp.all(pytree2[-1].bias == true_pytree2[-1].bias)
+
```
Comment (patrick-kidger) [inline tests/test_tree.py:49]: I think let's pull this out as a separate test, it's edge case enough to want to highlight things. Something like this is probably what we want I think:
```python
a = ()
b= (1,)
x1 = [a]
x2 = [a, a]
x3 = [(), ()]
x4 = [b]
x5 = [b, b]
x6 = [(1,), (1,)]

for x in (x1, x2, x3, x4, x5, x6):
    new_x = eqx.tree_at(lambda xi: xi[0], x, "hello")
    assert new_x[0] == "hello"
    if len(new_x) != 1:
        assert new_x[1] != "hello"
```

WDYT?

**Review by patrick-kidger — COMMENTED** [review]
Ah, awesome! Thank you for taking this on. I'm glad to see this doesn't look too tricky!

File: equinox/_tree.py (hunk 1 above)
Comment (lockwo) [reply equinox/_tree.py:67]: That's true, it was a hold over from my print debugging

File: equinox/_tree.py (hunk 2 above)
Comment (lockwo) [reply equinox/_tree.py]: done

File: tests/test_tree.py (hunk 3 above)
Comment (lockwo) [reply tests/test_tree.py]: Sure, makes sense, pulled it out

**Comment by patrick-kidger** [comment]
LGTM! Thank you for the fix here.

Indeed the latest JAX release broke some stuff, so in the meantime I'm going to merge this anyway.
